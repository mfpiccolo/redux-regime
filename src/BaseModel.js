import pluralize from "pluralize";
import QueryObject from "./QueryObject";

export default class BaseModel {
  static query(resources) {
    return new QueryObject(
      this,
      pluralize(this.name.toLowerCase()),
      resources,
      this.hasMany,
      this.belongsTo
    );
  }

  constructor(resources, attributes, hasMany = [], belongsTo = []) {
    Object.entries(attributes).forEach(([key, value]) => {
      this[key] = value;
    });

    if (hasMany.forEach) {
      hasMany.forEach(relationship =>
        this._buildHasManyQueryObject(this, resources, relationship)
      );
    }

    if (belongsTo.forEach) {
      belongsTo.forEach(relationship => {
        const relationshipKey = relationship.name.toLowerCase();
        this[relationshipKey] = () => {
          // needs to return the related model
        };
      });
    }
  }

  _filterResources(resource, resources, relationship, relationshipKey) {
    const currentResourceKey = pluralize(
      resource.constructor.name.toLowerCase()
    );
    const resourceClass = resource.constructor;
    const relationshipClass = relationship;
    return {
      ...resources,
      [currentResourceKey]: resources[currentResourceKey][resource.id],
      [relationshipKey]: relationshipClass
        .query(resources)
        .whereRelated(resourceClass, {
          id: resource.id
        }).currentResources
    };
  }

  _buildHasManyQueryObject(resource, resources, relationship) {
    const relationshipKey = pluralize(relationship.name.toLowerCase());
    if (!resource[relationshipKey]) {
      resource[relationshipKey] = () => {
        const newResouces = resource._filterResources(
          resource,
          resources,
          relationship,
          relationshipKey
        );

        return new QueryObject(
          relationship,
          relationshipKey,
          newResouces,
          relationship.hasMany,
          relationship.belongsTo
        );
      };
    }
  }
}
