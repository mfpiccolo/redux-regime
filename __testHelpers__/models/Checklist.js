import BaseQuery from "../../src/BaseQuery";
import Task from "./Task";

export default class Checklist extends BaseQuery {
  static get belongsTo() {
    return [];
  }

  static get hasMany() {
    return [Task];
  }
}
