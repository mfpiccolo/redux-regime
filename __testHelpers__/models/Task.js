import BaseQuery from "../../src/BaseQuery";
import Checklist from "./Checklist";

export default class Task extends BaseQuery {
  static get belongsTo() {
    return [Checklist];
  }

  randomMethod() {
    return "randomthing";
  }
}
