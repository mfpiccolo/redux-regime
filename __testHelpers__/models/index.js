import Checklist from "./Checklist";
import Task from "./Task";

export { Checklist, Task };
