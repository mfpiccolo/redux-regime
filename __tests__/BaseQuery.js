import BaseModel, { QueryObject } from "../src/BaseModel";

import resources
  from "../__testHelpers__/fixtrues/checklistsAndTasksNormalized";
import { Checklist, Task } from "../__testHelpers__/models";

describe("BaseModel", () => {
  test("query returns a QueryObject", () => {
    expect(BaseModel.query(resources)).toBeInstanceOf(QueryObject);
  });

  describe("models", () => {
    describe("all()", () => {
      test("", () => {
        expect(Checklist.query(resources).all().toModels()).toMatchSnapshot();
      });
    });

    describe("find()", () => {
      test("", () => {
        const checklist = Checklist.query(resources).find(1);
        expect(checklist.id).toEqual(1);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("where()", () => {
      test("filter by name", () => {
        const checklist = Checklist.query(resources)
          .where({ name: "Project Audit Rest" })
          .toModels()[0];
        expect(checklist.id).toEqual(2);
        expect(checklist).toMatchSnapshot();
      });

      test("filter by id", () => {
        const checklist = Checklist.query(resources)
          .where({ id: 3 })
          .toModels()[0];
        expect(checklist.id).toEqual(3);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("includes()", () => {
      test("", () => {
        const checklists = Checklist.query(resources)
          .all()
          .includes(["tasks"])
          .toModels();
        expect(checklists).toMatchSnapshot();
      });
    });

    describe("hasMany()", () => {
      test("hasMany", () => {
        const checklists = Checklist.query(resources)
          .find(1)
          .tasks()
          .toModels();
        expect(checklists).toMatchSnapshot();
      });
    });

    // describe("belongsTo()", () => {
    //   test("", () => {
    //     const checklist = Task.query(resources).checklist().toModels();
    //     expect(checklists).toMatchSnapshot();
    //   });
    // });
  });
  describe("objects", () => {
    describe("all()", () => {
      test("", () => {
        expect(Checklist.query(resources).all().toObjects()).toMatchSnapshot();
      });
    });

    describe("find()", () => {
      test("", () => {
        const checklist = Checklist.query(resources).find(1);
        expect(checklist.id).toEqual(1);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("where()", () => {
      test("", () => {
        const checklist = Checklist.query(resources)
          .where({ name: "Project Audit Rest" })
          .toObjects()[0];
        expect(checklist.id).toEqual(2);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("includes()", () => {
      test("", () => {
        const checklists = Checklist.query(resources)
          .all()
          .includes(["tasks"])
          .toObjects();
        expect(checklists).toMatchSnapshot();
      });
    });
  });
});
