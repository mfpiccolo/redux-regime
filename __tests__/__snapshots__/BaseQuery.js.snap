// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseQUery models all()  1`] = `
Array [
  Checklist {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": [Function],
  },
  Checklist {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": [Function],
  },
  Checklist {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": [Function],
  },
]
`;

exports[`BaseQUery models find()  1`] = `
Checklist {
  "id": 1,
  "name": "Onboarding Rest",
  "tasks": [Function],
}
`;

exports[`BaseQUery models hasMany() hasMany 1`] = `
Array [
  Task {
    "checklist": [Function],
    "description": "Onboarding REST 1",
    "id": 1,
  },
  Task {
    "checklist": [Function],
    "description": "Onboarding REST 2",
    "id": 2,
  },
]
`;

exports[`BaseQUery models includes()  1`] = `
Array [
  Checklist {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": Array [
      Task {
        "description": "Onboarding REST 1",
        "id": 1,
      },
      Task {
        "description": "Onboarding REST 2",
        "id": 2,
      },
    ],
  },
  Checklist {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": Array [
      Task {
        "description": "Project Audit Rest 3",
        "id": 3,
      },
      Task {
        "description": "Project Audit Rest 4",
        "id": 4,
      },
    ],
  },
  Checklist {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": Array [
      Task {
        "description": "QA Checklist Rest 5",
        "id": 5,
      },
      Task {
        "description": "QA Checklist Rest 6",
        "id": 6,
      },
    ],
  },
]
`;

exports[`BaseQUery models where() filter by id 1`] = `
Checklist {
  "id": 3,
  "name": "QA Checklist Rest",
  "tasks": [Function],
}
`;

exports[`BaseQUery models where() filter by name 1`] = `
Checklist {
  "id": 2,
  "name": "Project Audit Rest",
  "tasks": [Function],
}
`;

exports[`BaseQUery objects all()  1`] = `
Array [
  Object {
    "id": 1,
    "name": "Onboarding Rest",
  },
  Object {
    "id": 2,
    "name": "Project Audit Rest",
  },
  Object {
    "id": 3,
    "name": "QA Checklist Rest",
  },
]
`;

exports[`BaseQUery objects find()  1`] = `
Checklist {
  "id": 1,
  "name": "Onboarding Rest",
  "tasks": [Function],
}
`;

exports[`BaseQUery objects includes()  1`] = `
Array [
  Object {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": Array [
      Object {
        "description": "Onboarding REST 1",
        "id": 1,
      },
      Object {
        "description": "Onboarding REST 2",
        "id": 2,
      },
    ],
  },
  Object {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": Array [
      Object {
        "description": "Project Audit Rest 3",
        "id": 3,
      },
      Object {
        "description": "Project Audit Rest 4",
        "id": 4,
      },
    ],
  },
  Object {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": Array [
      Object {
        "description": "QA Checklist Rest 5",
        "id": 5,
      },
      Object {
        "description": "QA Checklist Rest 6",
        "id": 6,
      },
    ],
  },
]
`;

exports[`BaseQUery objects where()  1`] = `
Object {
  "id": 2,
  "name": "Project Audit Rest",
}
`;
